### PR TITLE
Upgrade Linux kernel to 4.14.336

### LINKED ISSUE #330: Kernel deblob script removes entire drivers for having even optional external firmware file loading support, including r8169 (most common Gigabit Ethernet chipsets)
Testing bare-metal bootstrapping on an Asus P5K Premium revealed that the r8169 Ethernet driver is being removed in its entirety by deblob. This driver contains no embedded blobs, only uses external firmware files optionally (works without them, just e.g. uses more power), and it's a requirement for Ethernet on the majority of consumer motherboards and NICs that would be used for bare-metal. (Luckily this board also has a Marvell Yukon Ethernet chip, which is unaffected.)

This was discovered during the workup of #329 (all Ethernet drivers disabled in config, except the ones used by qemu).

While deblobbing, in the sense of removing blobs _embedded_ in kernel driver code, is certainly a necessity to ensure the trustworthiness of the bootstrap environment, removing the firmware loading mechanism, and especially entire drivers that depend on or even just optionally use it, seems way too excessive - firmware loading alone can't compromise the environment, provided that the corresponding firmware files aren't present (which is already the case, since we don't include linux-firmware).

### PR DESCRIPTION
**Pull request by Googulator**
This is the last LTS version buildable using GCC 4.0.x. The next one, version 4.19, requires at least GCC 4.6.

Fortunately, this is also the first version of the Linux kernel without firmware blobs being included in /firmware, so the FSFLA
deblob scripts aren't needed anymore to ensure a fully auditable kernel - the 3 remaining drivers that do include blobs masquerading as source code are removed via a patch, avoiding all the other side effects of the deblob scripts.
This doesn't compromise the trustworthiness of the bootstrapped environment, since all the other drivers deblob would remove use the firmware loader mechanism, which does nothing when the actual firmware blobs aren't installed on the system separately. Features dependent on firmware still won't work, but many drivers that load firmware do so only optionally. This includes r8169, the driver for the Realtek gigabit NICs found on many x86 motherboards.

This kernel is considerably larger than 4.9.10, and we build more of it (including drivers that would previously get stripped away by
the deblob script, such as r8169), so to accommodate that, Fiwix initrd size is increased by 64MiB, while lowering kexec space by
the same amount to ensure enough userspace memory available in Fiwix.
Fiwix's maximum open file count is also bumped from 1.5K to 2.5K.

The Documentation folder is deleted before build, to further save space in the ramdisk.

Depends on #438 and https://github.com/oriansj/stage0-posix/pull/78

Fixes #330

### CODE REVIEW

**Review by stikonas — APPROVED** [review]

**Review by fosslinux — COMMENTED** [review]
Mostly LGTM, but I do want to review the config later today before merge

File: steps/linux-4.14.336/patches/ddfd117714014cebe7b5d008fdec8eaec620cba1.patch [hunk 1]
```diff
@@ -0,0 +1,60 @@
+SPDX-FileCopyrightText: 2024 Sean Christopherson <seanjc@google.com>
```
[the remark this answers is not in the document]
Comment (Googulator) [reply steps/linux-4.14.336/patches/ddfd117714014cebe7b5d008fdec8eaec620cba1.patch:1]: you're right, that's a typo

**Comment by Googulator** [comment]
> Config notes:
> 
>     * Different types of initramfs compression - do we need to care about anything other than gz?
The xz, lzo, etc. libraries still get included in the kernel for other purposes, so the space saving is minimal; on the flip side, keeping support here means we can in the future easily save RAM in Fiwix by decreasing the kexec buffer size and using higher compression for the initramfs.
>     * CONFIG_X86_BIGSMP=n. Should we enable this? Supports >8 "CPUs" which I think in kernel land might mean cores? Someone confirm please.
Nice catch, I was wondering why I can't set the max core count higher than 8, this is probably it.
>     * "Processor family" for optimisations appears to be set to "Pentium Pro" right now. Do we want to lower this? I think we'd technically support as low as 686 at the moment, so we should probably target that in the kernel.
"Pentium Pro" is Intel's trade name for a 686. The kernel offers no option for a "generic 686", as at that point, Intel's and AMD's implementations were diverging significantly.
I'd, however, consider upping the requirement (-march) to Pentium 3 and equivalent for AMD, with optimal tuning (-mtune) for Intel Core 2 and higher, since realistically that's as low as we can go, due to RAM requirements (i.e. no 686 can support 2GB+ RAM).
>     * I suspect we can disable all mouse support.
I don't think we need the savings from that right now, if it's even significant. I'd leave that for a future size reduction opportunity, if ever needed, especially since we already had mice enabled in 4.9.10.
>     * I think CONFIG_DRM_I915 can be disabled, that's an X-level driver I believe?
CONFIG_DRM_I915 (and other DRM drivers) isn't just for X; DRM drivers are the recommended way of getting a high resolution framebuffer on modern kernels, the plain-FB drivers are deprecated-ish and can even cause problems. In fact, I plan on switching to nouveau/radeon/amdgpu and disabling their FB counterparts, because nvidiafb turned out to be the cause of the green screen freeze bug on newer GeForce cards. (And yes, this functionality works without any firmware loaded; firmware is only required for 3D acceleration.)
This will, of course, be a followup PR, especially since it requires additional tricks to keep ramdisk usage under control.
>     * Maybe enable CONFIG_KVM, CONFIG_KVM_INTEL, CONFIG_KVM_AMD? I can imagine that one might like to run live-bootstrap within live-bootstrap on a bare metal system to verify they are the same.
KVM, unfortunately, doesn't compile with our current toolchain; I tried to patch it out, with no luck. It causes "undefined reference to `__compiletime_assert___COUNTER__'" errors at the final link, and unlike drm_edid, this doesn't seem to be patchable just by removing BUILD_BUG_ONs.

**Comment by Googulator** [comment]
With that, I think I will merge this now, and address the config issues in an upcoming 2nd PR, which should include the nvidiafb -> nouveau switch (so we don't lock up hard on newer GeForces, which nvidiafb unfortunately does), alongside other improvements.

**Pull request merged**